### ISSUE TITLE
Parser does not handle mixed case headers

### ISSUE THREAD

**Issue by jhwgh1968** [jedireza/warc#12]
I was testing an example WARC file from the much more developed [Perl CPAN Module](https://metacpan.org/pod/WARC), when I discovered that mixed-case headers are not currently supported.

When I try a file like this:

```
WARC/1.0
WARC-Type: warcinfo
Content-Type: text/plain
WARC-Date: 2019-09-02T23:25:55Z
WARC-Record-ID: <urn:test:file-1:record-0>
WARC-Filename: test-file-1.warc
Content-Length: 195

This is an example WARC file written by hand for testing purposes.  It may
or may not actually conform to any WARC specification, but is "close
enough" for testing the WARC library read support.
```

Then the `read_file` example fails:

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', examples/read_file.rs:16:45
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

That `None` is a `get()` on the headers. In other words, the parser did not find any headers at all!

I have an unpublished branch that would solve this problem, by turning the headers into an enum:

```
pub enum WarcHeader {
    CONTENT_LENGTH,
    CONTENT_TYPE,
    WARC_BLOCK_DIGEST,
    WARC_CONCURRENT_TO,
    WARC_DATE,
    WARC_FILENAME,
    WARC_IDENTIFIED_PAYLOAD_TYPE,
    WARC_IP_ADDRESS,
    WARC_PAYLOAD_DIGEST,
    WARC_PROFILE,
    WARC_RECORD_ID,
    WARC_REFERS_TO,
    WARC_SEGMENT_NUMBER,
    WARC_SEGMENT_ORIGIN_ID,
    WARC_SEGMENT_TOTAL_LENGTH,
    WARC_TARGET_URI,
    WARC_TRUNCATED,
    WARC_TYPE,
    WARC_WARCINFO_ID,
    Unknown(String),
}
```

This allows case insensitive parsing, while outputting lower case during generation (as it did before). It also validates certain headers are present, which are required by the format.

However, I recognize this is a significant API change, requiring a version `0.3.0` for sure. Once my gzip support is merged, would this be worth pursuing? A lot of tests would come with it, for sure.

**Comment by jedireza** [comment]
This is surprising. One of the tests in the parser uses mixed case headers and when I edit them (see diff) the tests still passes.

https://github.com/jedireza/warc/blob/5ed2c6bb25c81bd0a03f40d7946a1065920da74f/src/parser.rs#L198-L227

```diff
diff --git a/src/parser.rs b/src/parser.rs
index 8823710..1e97d0b 100644
--- a/src/parser.rs
+++ b/src/parser.rs
@@ -199,7 +199,7 @@ mod tests {
     fn parse_record() {
         let raw = b"\
             WARC/1.0\r\n\
-            Warc-Type: dunno\r\n\
+            WARC-Type: dunno\r\n\
             Content-Length: 5\r\n\
             \r\n\
             12345\r\n\
@@ -214,7 +214,7 @@ mod tests {

         let expected_version = "1.0";
         let expected_headers: Vec<(&str, &[u8])> =
-            vec![("Warc-Type", b"dunno"), ("Content-Length", b"5")];
+            vec![("WARC-Type", b"dunno"), ("Content-Length", b"5")];
         let expected_body: &[u8] = b"12345";

         assert_eq!(
```

But I have also confirmed that mixing the case in the example file headers and then running the read example does panic. 🤔

**Comment by jedireza** [comment]
🤦‍♂️ Now I see. This isn't a bug in the parser but in the fact that the header hash map tokens aren't normalized. We may want to simply convert them to lower case when parsing.

**Comment by jedireza** [comment]
```diff
diff --git a/src/warc_types.rs b/src/warc_types.rs
index 9a9f577..08e96d3 100644
--- a/src/warc_types.rs
+++ b/src/warc_types.rs
@@ -159,7 +159,7 @@ impl<R: BufRead> Iterator for WarcReader<R> {
             version: version_ref.to_owned(),
             headers: headers_ref
                 .into_iter()
-                .map(|(token, value)| (token.to_owned(), value.to_owned()))
+                .map(|(token, value)| (token.to_lowercase().to_owned(), value.to_owned()))
                 .collect(),
             body: body_ref.to_owned(),
         };
```

This fixes that particular issue. Thoughts?

**Comment by jedireza** [comment]
Resolves in #15

**Comment by jhwgh1968** [comment]
That solves the immediate problem, but I see value in an enum in other ways. I'll open a separate issue for that, though.

**Comment by jedireza** [comment]
We can keep this open so you don't need to create a new issue. I was just about to comment about this. I used to have much more elaborate code for headers: https://github.com/jedireza/warc/tree/8d69372fd2dc4b3c15dc3c45622675ae40511d5f/src/header

**Comment by jhwgh1968** [comment]
Oh! I didn't see your comment because I was too busy typing up the other issue. :laughing:

**Comment by jedireza** [comment]
No worries. I'm going to copy my comment over there.

**Comment by jhwgh1968** [comment]
In short, I wouldn't go that far, just something lightweight and simple defined within this crate in a single file. I already opened #16.
